### DISCUSSION TITLE
Feature request: support secrets for multiple PostgreSQL databases

### DISCUSSION THREAD

**Discussion by cdeil** [duckdb/duckdb#10963] [Ideas]
We have multiple PostgreSQL databases and I'd like to attach to them in DuckDB from my data pipeline and app.

At  https://duckdb.org/docs/extensions/postgres I see that I can use environment variables and pass PGPASSWORD which is nice and secure for one database (leaving the password out of my production code, but also e.g. out of my DuckDB shell for interactive work and thus any shell history and log).

Is this a new feature request?

Or already possible and could simply be explained to DuckDB noobs like me in the docs?

---

Current docs:

<img width="1001" alt="Screenshot 2024-03-03 at 14 23 02" src="https://github.com/duckdb/duckdb/assets/852409/8706a470-c405-4eb4-9027-b2e9333590a9">

**Comment by Tishj** [comment]
Can "also" be specified through environment variables
You can use the connection string instead when you have multiple databases to connect to :)

**Reply by cdeil** [reply]
Then what is best practice to manage multiple connection strings that contain the secrets? Is it possible with environment variables? Or what else is recommended?

**Reply by Tishj** [reply]
I dont think there is a way to do that then

**Comment by carlopi** [comment]
If you are using CLI you can make it work with the now deprecated postgres_attach combined with getenv:
```
carlo@ScroogeMcDuck duckdb % FOOD=pizza duckdb
v0.10.0 20b1486d11
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D CALL postgres_attach('dbname=postgres user=' || getenv('FOOD') || ' host=127.0.0.1');
Error: IO Error: Unable to connect to Postgres at dbname=postgres user=pizza host=127.0.0.1: connection to server at "127.0.0.1", port 5432 failed: Connection refused
	Is the server running on that host and accepting TCP/IP connections?
```
In python / other bindings you can probably do the same using the python equivalent to get the env variable.

The proper way would be using [secrets](https://duckdb.org/docs/configuration/secrets_manager) to pass credentials to attach, but this has not been shipped yet.

**Reply by dpprdan** [reply]
> The proper way would be using [secrets](https://duckdb.org/docs/configuration/secrets_manager) to pass credentials to attach, but this has not been shipped yet.

See the page you linked to:

> Secrets were introduced with DuckDB version 0.10. 

Or did you mean Secrets for PostgreSQL? Because Secrets with `TYPE POSTGRESQL` are not supported (yet?) (`Error: Invalid Input Error: Secret type 'postgresql' not found`) and AFAICT there is no open or recently merged PR that would add support for it.

(I use capitalized "Secrets" for the duckdb feature and "secrets" as meaning credentials in general).

Also, AFAICT Secrets does not support `getenv()` (yet?) so it wouldn't really help e.g. in a CI setting where you'd want to pass your connection settings with (non-standard) environment variables.

**Reply by carlopi** [reply]
DuckDB's secret manager & the postgres scanner will not work right now, as hinted by the "this has not been shipped yet" that I though was clear enough.



will not handle postgres credentials in the current form,

**Reply by dpprdan** [reply]
> that I though was clear enough.

Well, "this" in "this has not been shipped yet" can refer to "using [secrets](https://duckdb.org/docs/configuration/secrets_manager)" in general or to "using [secrets](https://duckdb.org/docs/configuration/secrets_manager) to pass credentials to attach", hence my confusion. 

Is there an issue or PR where this work is tracked? ["secrets postgres" comes up empty](https://github.com/duckdb/duckdb/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+secrets+postges), as does "[secrets getenv](https://github.com/duckdb/duckdb/issues?q=sort%3Aupdated-desc+secrets+getenv)" (which are separate issues IMHO).

**Reply by cdeil** [reply]
@carlopi - thanks!

Well, then please consider this thread a feature request to improve env var and/or secrets management in DuckDB to enable multiple PostgreSQL database connections (or more generally attach to multiple databases of a given type)

**Reply by dpprdan** [reply]
@cdeil You may want put "Feature Request" in the title and change the category to "ideas", though I am not sure whether that makes a difference. 

> improve env var and/or secrets management in DuckDB to [...] attach to multiple databases of a given type

(Kind of repeating myself here, but) I think this means adding Postgres (MySQL/MariaDB, etc.) `TYPE`s  of Secrets and a way to pass environment variables (and possibly config files like .pg_service.conf/.pgpass) to them. At least that's what I gather from this discussion and the general idea of Secrets being the way forward.

Pinging @samansmink since I spotted Postgres under Followup in [his Secrets PR](https://github.com/duckdb/duckdb/pull/10042) 🙈

**Comment by dpprdan** [comment]
I have the same or slightly more narrowed down question: How can you pass Postgres connection parameters as (non-standard) environment variables to the duckdb CLI?

With "non-standard" I mean different from the [libpq env vars](https://www.postgresql.org/docs/current/libpq-envars.html) (PGUSER, PGPASSWORD, etc). 

In my use case I also use multiple Postgres Databases with different credentials within the same duckdb session (i.e. using duckdb as a poor-mans ETL tool) within a CI where I do not have (full) control over the environment variables.

My naive attempts so far have been: 

1. Possibly the most naive assumption that `getenv()` gets resolved from the connection string:

```
ATTACH 'service=myservice user=getenv(''MYUSER'') password=getenv(''MYPASSWORD'')' AS mypg (TYPE postgres);
```

results in 

```
[...] no pg_hba.conf entry for host "84.158.161.85", user "getenv('MYUSER')" [...]
```

(BTW I use a [Connection Service File](https://www.postgresql.org/docs/current/libpq-pgservice.html) (i.e. `.pg_service.conf`) for all connection parameters except the user credentials. I am aware of `.pgpass` but cannot submit this as an environment variable to CI and do not want to commit it to the repo obviously).

2. `concat()` a connection string:

```
ATTACH concat('service=myservice', ' user=', getenv('MYUSER'), ' password=', getenv('MYPASSWORD')) AS mypg (TYPE postgres);
```

results in

```
Error: Parser Error: syntax error at or near "concat"
LINE 1: ATTACH concat('service=myservice', ' user=', getenv(...
```

3. set environment variables to the libpq ones from within duckdb (e.g. before connecting to the [second DB](https://github.com/duckdb/duckdb/discussions/4460#discussioncomment-3444851)).

```
SET PGUSER = getenv('MYUSER');
```

```
Error: Catalog Error: unrecognized configuration parameter "PGUSER"

Did you mean: "user", "pg_use_binary_copy", "username", "preserve_identifier_case", "preserve_insertion_order"
```

4. Use the [Secrets Manager](https://duckdb.org/docs/configuration/secrets_manager) (see my reply to the other answer regarding Secrets).

**Reply by carlopi** [reply]
You should try using the currently deprecated: `CALL postgres_attach('dbname=postgres user=' || getenv('FOOD') || ' host=127.0.0.1');`, see my answer, that might allow a way to concatenate gevenv() calls to the string.

**Reply by dpprdan** [reply]
This sort-of works, but like you say, it is deprecated. E.g. it doesn't add a [catalog](https://duckdb.org/docs/sql/statements/attach.html#name-qualification) for the attached database (AFAICT) which is problematic if you want to attach multiple databases. I guess you could also work around that (e.g. `ATTACH` multiple in-memory duckdb instances and `postgres_attach` not more than one Postgres DBs to each of those?), but all in all nothing I would want to start a new project with.

**Reply by carlopi** [reply]
Unsure on the way forward, maybe someone else can join the conversation.

**Comment by mehd-io** [comment]
Note that secrets manager now supports Postgres/MySQL since September release: https://duckdb.org/docs/extensions/postgres#configuring-via-secrets
You can also specify which secrets to use : 
```
ATTACH '' AS postgres_db_one (TYPE POSTGRES, SECRET postgres_secret_one);
```

**Reply by dmcmurchy** [reply]
@mehd-io - The password isn't kept secret (redacted) in all locations. I can elaborate but wasn't sure if this was considered a security vulnerability or just a bug?

**Comment by vfilter** [comment]
Just to clarify: The original request is not addressed, correct? 
I was imaging to be able to just pass a complete postgres connection string as an env variable into duckdb and use it when attaching a postgres database, but that doesn't seem possible
